lib/synapse/haproxy.rb

```diff
@@ -655,18 +693,33 @@ def generate_frontend_stanza(watcher, config)
     end
 
     def generate_backend_stanza(watcher, config)
+      backends = {}
+
+      # The ordering here is important.  First we add all the backends in the
+      # disabled state...
+      @seen.fetch(watcher.name, []).each do |backend_name, backend|
+        backends[backend_name] = backend.merge('enabled' => false)
+      end
+
+      # ... and then we overwite any backends that the watchers know about,
+      # setting the enabled state.
+      watcher.backends.each do |backend|
+        backend_name = construct_name(backend)
+        backends[backend_name] = backend.merge('enabled' => true)
+      end
+
       if watcher.backends.empty?
-        log.warn "synapse: no backends found for watcher #{watcher.name}"
+        log.debug "synapse: no backends found for watcher #{watcher.name}"
       end
 
       stanza = [
         "\nbackend #{watcher.name}",
         config.map {|c| "\t#{c}"},
-        watcher.backends.shuffle.map {|backend|
-          backend_name = construct_name(backend)
+        backends.map {|backend_name, backend|
```

Comment:
@schleyfox It doesn't really work nicely, it works as just hashing. 

In order to work better with consistent hashing we may keep dead backends for longer.

This may be done by creating two keys in zk, one listing all the machines in the pool and the other telling if it's alive. The key listing all the machines may have an adjustable timeout and the default value should be low to work nicely everywhere.

However in case the ordering is really important, we may also introduce sequential numbering of nodes and if an hole is detected in the middle we may fill it with a fake machine. This fake machine will be replaced once a new one joins the cluster.

Btw I'm ok with making shuffling the default option, as long as this is a very specific use case.

OT: take a look at https://github.com/dockerparis/DockStack, it makes really easy to use nerve/synapse as a Docker ambassador.